### PR TITLE
Add helper for logit attribution

### LINKED ISSUE #109: Add helper function for direct logit attribution
Add a helper function (eg to `ActivationCache`) to calculate direct logit attribution easily. Crib from the code here: https://colab.research.google.com/github/neelnanda-io/Easy-Transformer/blob/main/Exploratory_Analysis_Demo.ipynb#scrollTo=Direct_Logit_Attribution

### PR DESCRIPTION
**Pull request by dkamm**
Closes #109 

Hi @neelnanda-io - I took at stab at this (basically just lifted the exploratory demo code).  Let me know if this is what you had in mind

I made `residual_stack` a parameter since it's doing the same thing for logit lens, per layer and per head attributions in the exploratory demo.  The code expects `residual_stack` to have a batch dimension since I apply layer norm inside.

The code supports `int` and `str` values for `tokens` and `input_tokens` since that seems to be an idiom but you'll have to pass in `residual_stack[:,i,:].unsqueeze(dim=1)` to get the shapes right.  Not sure how important this is.

It should also support `tokens` and `input_tokens` with shapes like `[batch, position]` (as opposed to just `[batch]`), but I'm not sure what that would be used for? 

The code takes the average over the batch, not sure if this should be configurable

Re: testing - I can try adding some unit tests but it doesn't seem trivial to set up and might be better as a separate PR for #99 

Example usage:
```python
# logit lens attributions
accumulated_residual, labels = cache.accumulated_resid(layer=-1, incl_mid=True, pos_slice=-1, return_labels=True)
logit_lens_logit_diffs = cache.logit_attrs(residual_stack=accumulated_residual, tokens=tokens, incorrect_tokens=incorrect_tokens, layer=-1, pos_slice=-1)

# per head attributions
per_head_residual, labels = cache.stack_head_results(layer=-1, pos_slice=-1, return_labels=True)
per_head_logit_diffs = cache.logit_attrs(residual_stack=per_head_residual, tokens=tokens, incorrect_tokens=incorrect_tokens, layer=-1, pos_slice=-1)
per_head_logit_diffs = einops.rearrange(per_head_logit_diffs, "(layer head_index) -> layer head_index", layer=model.cfg.n_layers, head_index=model.cfg.n_heads)

# single token attribution (batch size = 1)
accumulated_residual_single = accumulated_residual[:,0,:].unsqueeze(dim=1)
logit_lens_logit_diff_single = cache.logit_attrs(residual_stack=accumulated_residual_single, tokens=tokens[0], incorrect_tokens=incorrect_tokens[0], layer=-1, pos_slice=-1)
```

### CODE REVIEW

**Review by neelnanda-io — DISMISSED** [review]
Thanks for the PR! I think I vibe with the overall design, but want to significantly change the handling of tokens (put everything into a tensor of consistent shape), and to allow batch, position or both dimensions for the residual stack, and to not average over the batch in the output.

File: transformer_lens/ActivationCache.py [hunk 1]
```diff
@@ -174,6 +174,67 @@ def accumulated_resid(
         else:
             return components
 
+    def logit_attrs(
+            self,
+            residual_stack: TT["components", "batch", "d_model"],
+            tokens: Union[str, int, TT[()], TT["batch"], TT["batch", "position"]],
+            incorrect_tokens: Optional[Union[str, int, TT[()], TT["batch"], TT["batch", "position"]]],
+            layer: Optional[int] = None,
```
[the remark this answers is not in the document]
Comment (dkamm) [reply transformer_lens/ActivationCache.py:182]: Makes sense!

File: transformer_lens/ActivationCache.py [hunk 2]
```diff
@@ -174,6 +174,67 @@ def accumulated_resid(
...
+            pos_slice (Slice): The slice to apply layer norm scaling on.
+                Defaults to None, do nothing.
+
+        Returns:
+            Components: A [num_components] or [num_components, position] tensor of the logit attributions averaged across the batch. 
+        """
+        if not isinstance(pos_slice, Slice):
+            pos_slice = Slice(pos_slice)
+
+        tokens_is_tensor = isinstance(tokens, torch.Tensor) and len(tokens.shape) > 0
```
[the remark this answers is not in the document]
Comment (dkamm) [reply transformer_lens/ActivationCache.py:202]: Agree that is a much better way! What I had was pretty silly lol

File: transformer_lens/ActivationCache.py [hunk 3]
```diff
@@ -174,6 +174,67 @@ def accumulated_resid(
...
+            # incorrect_tokens is scalar, tokens is tensor
+            raise ValueError(f"tokens and incorrect_tokens have incompatible shapes- check these arguments? (incorrect_tokens={incorrect_tokens}, tokens.shape={tokens.shape})")
+
+        logit_directions = self.model.tokens_to_residual_directions(tokens)
+        if incorrect_tokens is not None:
+            # If incorrect_tokens was provided, take the logit difference
+            logit_directions -= self.model.tokens_to_residual_directions(incorrect_tokens)
+
+        if not tokens_is_tensor:
+            # Add dummy dimension for batch since we expect residual_stack to have a batch dimension
```
[the remark this answers is not in the document]
Comment (dkamm) [reply transformer_lens/ActivationCache.py:231]: Short answer: I realize now it doesn't so feel free to move on lol

Long answer (if curious): 
So this came from a misunderstanding that arose from how I most likely misused the library

I wanted to test on a single example with batch dimension removed and did the following:
```python
# tokens.shape: [8, 15]
_, cache = model.run_with_cache(tokens)
accumulated_residual = cache.accumulated_resid(layer=-1, incl_mid=True, pos_slice=-1)
resid_stack = accumulated_residual[:,0,:] # I tried to apply ln on a single example with batch dimension removed
scaled_resid_stack = cache.apply_ln_to_stack(resid_stack, layer=-1, pos_slice=-1)
# Throws error
```
I misinterpreted this error as residual stack needing to have a batch dimension.  Upon closer inspection, I see that scale is cached for the whole batch and that's where the shape mismatch came from.

I think right way to test what I wanted is rerun the model on a single example with no batch dimension?
```python
# tokens.shape: [8, 15]
token = tokens[:,0] # take first example
_, cache = model.run_with_cache(token)
accumulated_residual = cache.accumulated_resid(layer=-1, incl_mid=True, pos_slice=-1)
resid_stack = accumulated_residual.squeeze() # accumulated residual shape is [25, 1, 768], so squeeze to remove batch dim
scaled_resid_stack = cache.apply_ln_to_stack(resid_stack, layer=-1, pos_slice=-1)
# works
```

File: transformer_lens/ActivationCache.py [hunk 4]
```diff
@@ -174,6 +174,67 @@ def accumulated_resid(
...
+            residual_stack (TT["components", "batch", "d_model"]): stack of components of residual stream to get logit attributions for.
+            tokens (Union[str, int, TT[()], TT["batch"], TT["batch", "position"]]): tokens to compute logit attributions on.
+            incorrect_tokens (Optional[Union[str, int, TT[()], TT["batch"], TT["batch", "position"]]]): if provided, compute attributions on logit difference between tokens and incorrect_tokens.
+            layer (int): The layer to use as input for layer norm scaling.
+                Defaults to None, which maps to the n_layers case, ie the unembed.
+            pos_slice (Slice): The slice to apply layer norm scaling on.
+                Defaults to None, do nothing.
+
+        Returns:
+            Components: A [num_components] or [num_components, position] tensor of the logit attributions averaged across the batch.
```
[the remark this answers is not in the document]
Comment (dkamm) [reply transformer_lens/ActivationCache.py:197]: Agree! It's pretty trivial to do `logit_attrs.mean(dim=-1)` outside

File: transformer_lens/ActivationCache.py [hunk 5]
```diff
@@ -174,6 +174,67 @@ def accumulated_resid(
...
+                raise ValueError(f"tokens and incorrect_tokens have incompatible shapes- check these arguments? (incorrect_tokens.shape={incorrect_tokens.shape}, tokens.shape={tokens.shape})")
+        
+        elif incorrect_tokens and tokens_is_tensor:
+            # incorrect_tokens is scalar, tokens is tensor
+            raise ValueError(f"tokens and incorrect_tokens have incompatible shapes- check these arguments? (incorrect_tokens={incorrect_tokens}, tokens.shape={tokens.shape})")
+
+        logit_directions = self.model.tokens_to_residual_directions(tokens)
+        if incorrect_tokens is not None:
+            # If incorrect_tokens was provided, take the logit difference
+            logit_directions -= self.model.tokens_to_residual_directions(incorrect_tokens)
```
[the remark this answers is not in the document]
Comment (dkamm) [reply transformer_lens/ActivationCache.py:228]: Got it!

**Comment by dkamm** [comment]
Thanks for your patience and thorough comments!  I've made the changes and added my replies

Here's the script I used to test 

```python

# Janky code to do different setup when run in a Colab notebook vs VSCode
DEBUG_MODE = False
try:
    import google.colab
    IN_COLAB = True
    print("Running as a Colab notebook")
    %pip install git+https://github.com/neelnanda-io/TransformerLens.git
    # Install another version of node that makes PySvelte work way faster
    !curl -fsSL https://deb.nodesource.com/setup_16.x | sudo -E bash -; sudo apt-get install -y nodejs
    %pip install git+https://github.com/neelnanda-io/PySvelte.git
except:
    IN_COLAB = False
    print("Running as a Jupyter notebook - intended for development only!")
    from IPython import get_ipython

    ipython = get_ipython()
    # Code to automatically update the HookedTransformer code as its edited without restarting the kernel
    ipython.magic("load_ext autoreload")
    ipython.magic("autoreload 2")

# Import stuff
import torch
from fancy_einsum import einsum
from transformer_lens import HookedTransformer, HookedTransformerConfig, FactoredMatrix, ActivationCache
from torchtyping import TensorType as TT

# Turn off auto diff to save memory since only doing inference
torch.set_grad_enabled(False)

# Load gpt2-small
model = HookedTransformer.from_pretrained(
    "gpt2-small",
    center_unembed=True,
    center_writing_weights=True,
    fold_ln=True,
    refactor_factored_attn_matrices=True,
)

# Create IOI prompts
prompt_format = [
    "When John and Mary went to the shops,{} gave the bag to",
    "When Tom and James went to the park,{} gave the ball to",
    "When Dan and Sid went to the shops,{} gave an apple to",
    "After Martin and Amy went to the park,{} gave a drink to",
]
names = [
    (" Mary", " John"),
    (" Tom", " James"),
    (" Dan", " Sid"),
    (" Martin", " Amy"),
]
# List of prompts
prompts = []
# List of answers, in the format (correct, incorrect)
answers = []
# List of the token (ie an integer) corresponding to each answer, in the format (correct_token, incorrect_token)
answer_tokens = []
for i in range(len(prompt_format)):
    for j in range(2):
        answers.append((names[i][j], names[i][1 - j]))
        answer_tokens.append(
            (
                model.to_single_token(answers[-1][0]),
                model.to_single_token(answers[-1][1]),
            )
        )
        # Insert the *incorrect* answer to the prompt, making the correct answer the indirect object.
        prompts.append(prompt_format[i].format(answers[-1][1]))
answer_tokens = torch.tensor(answer_tokens)

tokens = model.to_tokens(prompts, prepend_bos=True)

# Run the model and cache all activations
original_logits, cache = model.run_with_cache(tokens)

# Reference notebook code
answer_residual_directions = model.tokens_to_residual_directions(answer_tokens)
logit_diff_directions = answer_residual_directions[:, 0] - answer_residual_directions[:, 1]

def residual_stack_to_logit_diff(residual_stack: TT["components", "batch", "d_model"], cache: ActivationCache) -> float:
    scaled_residual_stack = cache.apply_ln_to_stack(residual_stack, layer = -1, pos_slice=-1)
    return einsum("... batch d_model, batch d_model -> ...", scaled_residual_stack, logit_diff_directions)/len(prompts)

# Get accumulated resid
accumulated_residual, labels = cache.accumulated_resid(layer=-1, incl_mid=True, pos_slice=-1, return_labels=True)

# Get reference logit diffs
ref_ave_logit_diffs = residual_stack_to_logit_diff(accumulated_residual, cache)

# Get our logit diffs
logit_diffs = cache.logit_attrs(accumulated_residual, pos_slice=-1, tokens=answer_tokens[:,0], incorrect_tokens=answer_tokens[:,1])
ave_logit_diffs = logit_diffs.mean(dim=-1)

# Print difference between ref and ours
print("Ref vs ours difference: ", (ref_ave_logit_diffs - ave_logit_diffs).abs().sum())

# Ref vs ours difference:  tensor(2.5898e-06)
```

File: transformer_lens/ActivationCache.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (neelnanda-io) [reply transformer_lens/ActivationCache.py]: Ah! Yes, that is correct. But also, this is probably a bug in apply_ln_to_stack, and we should include a param for `batch_slice`, since this is an obviously reasonable use case

[the remark this answers is not in the document]
Comment (neelnanda-io) [reply transformer_lens/ActivationCache.py]: Feel free to include that in this PR

[the remark this answers is not in the document]
Comment (dkamm) [reply transformer_lens/ActivationCache.py]: Hey Neel - I added the `batch_slice` param to both functions.

I try to determine if the residual stack has a batch dimension in `apply_ln_to_stack`.  I think it works because you should be passing the same `pos_slice` that you applied on the residual stack, but might want to double check this.

Can do these now

```
# All tokens
log_diffs = cache.logit_attrs(accumulated_residual, pos_slice=-1, tokens=answer_tokens[:,0], incorrect_tokens=answer_tokens[:,1]) # [25, 8, 758]

# Single token
batch = -1
log_diffs = cache.logit_attrs(accumulated_residual, batch_slice=batch, pos_slice=-1, tokens=answer_tokens[batch,0], incorrect_tokens=answer_tokens[batch,1]) # [25, 758]

# Single token (batchless residual stack)
batch = -1
log_diffs = cache.logit_attrs(accumulated_residual[:,batch,:], pos_slice=-1, tokens=answer_tokens[batch,0], incorrect_tokens=answer_tokens[batch,1]) # [25, 758]

# Array slice of tokens
batch = [2,5,7]
log_diffs = cache.logit_attrs(accumulated_residual, batch_slice=batch, pos_slice=-1, tokens=answer_tokens[batch,0], incorrect_tokens=answer_tokens[batch,1]) # [25, 3, 758]
```

**Comment by neelnanda-io** [comment]
Thanks for the changes! Other than my comment about the len(scale.shape)>1 step, this LGTM

**Comment by neelnanda-io** [comment]
And can you add your test script as a test? It'd be better to have actual ActivationCache tests, but this seems better than nothing!

**Comment by neelnanda-io** [comment]
Ideally on eg solu-2l, or some other tiny model, to make tests faster

**Comment by dkamm** [comment]
Sure, I'll add the script! Think it makes sense to work on #99 next lol

**Comment by dkamm** [comment]
@neelnanda-io this should be ready now 🤞 sorry for all the back and forth!

Added another test which revealed some bugs, which are now fixed

I also ran into a CI failure but couldn't reproduce on my machine- bumping the atol to 1e-7 got rid of it which I think is ok

Also please squash this when merging 😭

**Comment by jbloomAus** [comment]
Attempted to fix merge conflict, looked very simple (torch vs Jaxtyping)

**Review by jbloomAus — APPROVED** [review]
changes are good!
